Add SKUs table, use newer API to fetch data

simark opened the pull request:
We currently don't handle multi-SKU products well, fix that.

TLDR:

 - Add database table for SKUs
 - Use newer API to be able to get more details about SKUs
 - Adjust web UI to account for that

Taking this product as an example:

https://www.canadiantire.ca/en/pdp/thule-convoy-xt-rooftop-cargo-box-0401588p.html

This product has 3 different variants, or SKUs, with three different
prices.  There is a single entry in the products_static table.  The
`sku` columns contains a list of all the three SKUs for that product:

  40-1588-0|40-1589-8|40-1590-2

In the products_dynamic table, the samples for multi-SKU products refer
to the complete product, not a specific SKU.  For the example product,
we have samples like this one:

    {
        "Banner": "CTR",
        "Product": "0401588P",
        "PrimarySKU": "0401588",
        "CheckDigit": "0",
        "SKU_Count": 3,
        "Price": 699.99,
        "PriceFrom": "Y",
        "Description": "THL ALP 12CFT CNV XT",
        "Messages": {
            "Warranty": "This product carries a 1 year exchange warranty redeemable at any Canadian Tire store."
        },
        "PartNumber": "632100",
        "IsOnline": {
            "Active": "Y",
            "Exclusive": "N",
            "Sellable": "Y",
            "SpecialBuy": "N"
        }
    }

Note "PriceFrom" is "Y".  I believe this means that the rendered product
list would say "From $699.99", meaning that $699.99 is the price of one
of the three SKUs, we don't know which one (whichever is the
cheapest).  Given that we want to track prices for each individual SKU,
samples like this one are not helpful, and will have to be discarded.

Samples that refer to a single-SKU product have a `SKU` key, so we can
be confident the price is for a specific SKU.

Changes at the database level:

  - add the `skus` table, with the following columns:
    - index: auto-incremented integer
    - code: code, in the format "0401589"
    - formatted_code: code, in the format "040-1589-8"
    - product_index: index of the product (products_static table) this
      SKU belongs to
  - remove the `skus` column of the `products_static` table.
  - add the `samples` table, which is like the existing
    `products_dynamic` table, but with a foreign reference to the `skus`
    table, instead of to the `products_static` table .
  - remove the `products_dynamic` table.

The migration does the following steps to keep as much of the existing
data as possible:

  - Populate the `skus` table based on the existing content of the `sku`
    column of the `products_static` table, whose format is shown above.
  - Go over each existing price sample (`products_dynamic` table), see
    if we can move it to the new table.  If it doesn't have a `SKU` key,
    or has 'PriceFrom: Y' (I think these two are correlated, but we
    check both just in case), we can't.  Use the `SKU` value to know
    which row in the `skus` table this sample will now refer to.
  - There are some products without SKU information in column
    `products_static.sku`.  This is probably because these products were
    not seen since we added the `sku` column.  For those, we did not
    create an entry in the `skus` table.  If some samples refer to that
    SKU, we have a problem.  In that case, create a `skus` row based on
    that SKU number.  Assume that the product code is the SKU code plus
    'P', which always seems to be the case for single-SKU products.
    There are some rare cases where there is no such product, in that
    case we drop the sample.

Adjust the scraping code to provide this new information.  To do so, use
newer APIs, which are those I see the Canadian Tire website use at the
time of writing.  In the new API used for fetching list of products, the
information about SKUs is better structured than in the previous one.

We previously limited ourselves to scraping the categories of level 1
and 2.  During my testing, I saw that scraping each level of category
gave a different number of items, and there always seemed to be items
that appeared only in one level and not the others.  To have the best
coverage, we would have to scrape all category levels.  But doing so
would increase the workload quite a bit, if done each day.  The tradeoff
I made to have good coverage but not a huge workload is to use one
category level each day.  The category level is based on the number of
the current day, so if the scraping is done every day (which is the case
in our production system), we'll do all category levels in a round robin
fashion.  This step is to discover new products/skus, so it's fine if we
don't discover a new one immediately.  This is only the default though,
the category levels to scrape can be specified on the command-line.

I removed the step where we fetch the number of items in each category
to provide a length to ProductLedger.  When scraping the level 5 (the
most specific categories), this is just too long, since there are
hundreds of categories.  As a consequence, we no longer show a progress
bar with a known length when scraping the inventory.  We could maybe add
it back later, and make it optional.

Adjust the web UI to add a "SKU" page.  So, the user lands on the index,
which is the list of products.  The user clicks on a product, they land
on the list of SKUs for that product.  They then click on the SKU they
want.  The experience sucks, it will have to be improved, but it works.
The existing product.html page becomes sku.html, with minor obvious
changes (but git won't tell you that).

Adjust the ctquery command accordingly.

simark committed:
Add SKUs table, use newer API to fetch data

We currently don't handle multi-SKU products well, fix that.

TLDR:

 - Add database table for SKUs
 - Use newer API to be able to get more details about SKUs
 - Adjust web UI to account for that

Taking this product as an example:

https://www.canadiantire.ca/en/pdp/thule-convoy-xt-rooftop-cargo-box-0401588p.html

This product has 3 different variants, or SKUs, with three different
prices.  There is a single entry in the products_static table.  The
`sku` columns contains a list of all the three SKUs for that product:

  40-1588-0|40-1589-8|40-1590-2

In the products_dynamic table, the samples for multi-SKU products refer
to the complete product, not a specific SKU.  For the example product,
we have samples like this one:

    {
        "Banner": "CTR",
        "Product": "0401588P",
        "PrimarySKU": "0401588",
        "CheckDigit": "0",
        "SKU_Count": 3,
        "Price": 699.99,
        "PriceFrom": "Y",
        "Description": "THL ALP 12CFT CNV XT",
        "Messages": {
            "Warranty": "This product carries a 1 year exchange warranty redeemable at any Canadian Tire store."
        },
        "PartNumber": "632100",
        "IsOnline": {
            "Active": "Y",
            "Exclusive": "N",
            "Sellable": "Y",
            "SpecialBuy": "N"
        }
    }

Note "PriceFrom" is "Y".  I believe this means that the rendered product
list would say "From $699.99", meaning that $699.99 is the price of one
of the three SKUs, we don't know which one (whichever is the
cheapest).  Given that we want to track prices for each individual SKU,
samples like this one are not helpful, and will have to be discarded.

Samples that refer to a single-SKU product have a `SKU` key, so we can
be confident the price is for a specific SKU.

Changes at the database level:

  - add the `skus` table, with the following columns:
    - index: auto-incremented integer
    - code: code, in the format "0401589"
    - formatted_code: code, in the format "040-1589-8"
    - product_index: index of the product (products_static table) this
      SKU belongs to
  - remove the `skus` column of the `products_static` table.
  - add the `samples` table, which is like the existing
    `products_dynamic` table, but with a foreign reference to the `skus`
    table, instead of to the `products_static` table .
  - remove the `products_dynamic` table.

The migration does the following steps to keep as much of the existing
data as possible:

  - Populate the `skus` table based on the existing content of the `sku`
    column of the `products_static` table, whose format is shown above.
  - Go over each existing price sample (`products_dynamic` table), see
    if we can move it to the new table.  If it doesn't have a `SKU` key,
    or has 'PriceFrom: Y' (I think these two are correlated, but we
    check both just in case), we can't.  Use the `SKU` value to know
    which row in the `skus` table this sample will now refer to.
  - There are some products without SKU information in column
    `products_static.sku`.  This is probably because these products were
    not seen since we added the `sku` column.  For those, we did not
    create an entry in the `skus` table.  If some samples refer to that
    SKU, we have a problem.  In that case, create a `skus` row based on
    that SKU number.  Assume that the product code is the SKU code plus
    'P', which always seems to be the case for single-SKU products.
    There are some rare cases where there is no such product, in that
    case we drop the sample.

Adjust the scraping code to provide this new information.  To do so, use
newer APIs, which are those I see the Canadian Tire website use at the
time of writing.  In the new API used for fetching list of products, the
information about SKUs is better structured than in the previous one.

We previously limited ourselves to scraping the categories of level 1
and 2.  During my testing, I saw that scraping each level of category
gave a different number of items, and there always seemed to be items
that appeared only in one level and not the others.  To have the best
coverage, we would have to scrape all category levels.  But doing so
would increase the workload quite a bit, if done each day.  The tradeoff
I made to have good coverage but not a huge workload is to use one
category level each day.  The category level is based on the number of
the current day, so if the scraping is done every day (which is the case
in our production system), we'll do all category levels in a round robin
fashion.  This step is to discover new products/skus, so it's fine if we
don't discover a new one immediately.  This is only the default though,
the category levels to scrape can be specified on the command-line.

I removed the step where we fetch the number of items in each category
to provide a length to ProductLedger.  When scraping the level 5 (the
most specific categories), this is just too long, since there are
hundreds of categories.  As a consequence, we no longer show a progress
bar with a known length when scraping the inventory.  We could maybe add
it back later, and make it optional.

Adjust the web UI to add a "SKU" page.  So, the user lands on the index,
which is the list of products.  The user clicks on a product, they land
on the list of SKUs for that product.  They then click on the SKU they
want.  The experience sucks, it will have to be improved, but it works.
The existing product.html page becomes sku.html, with minor obvious
changes (but git won't tell you that).

Adjust the ctquery command accordingly.

jgalar commented:
Really sorry for not looking at this earlier.

I'm getting this error during the migration:
```
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 4e2a96338a6c -> 9169a7c5bda3, add skus table
>>> Reading product list, generating SKU list
>>> Creating skus table
Traceback (most recent call last):
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1782, in _execute_context
    self.dialect.do_executemany(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 729, in do_executemany
    cursor.executemany(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: skus.code

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/config.py", line 588, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/config.py", line 582, in main
    self.run_cmd(cfg, options)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/config.py", line 559, in run_cmd
    fn(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/command.py", line 320, in upgrade
    script.run_env()
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/script/base.py", line 563, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/util/pyfiles.py", line 92, in load_python_file
    module = load_module_py(module_id, path)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/util/pyfiles.py", line 108, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/jgalar/Projects/CanadianTracker/alembic/env.py", line 84, in <module>
    run_migrations_online()
  File "/home/jgalar/Projects/CanadianTracker/alembic/env.py", line 78, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/runtime/environment.py", line 851, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/runtime/migration.py", line 620, in run_migrations
    step.migration_fn(**kw)
  File "/home/jgalar/Projects/CanadianTracker/alembic/versions/9169a7c5bda3_add_skus_table.py", line 113, in upgrade
    op.bulk_insert(skus_table, sku_values)
  File "<string>", line 8, in bulk_insert
  File "<string>", line 3, in bulk_insert
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/operations/ops.py", line 2310, in bulk_insert
    operations.invoke(op)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/operations/base.py", line 392, in invoke
    return fn(self, operation)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/operations/toimpl.py", line 198, in bulk_insert
    operations.impl.bulk_insert(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/ddl/impl.py", line 424, in bulk_insert
    self._exec(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/alembic/ddl/impl.py", line 193, in _exec
    return conn.execute(construct, multiparams)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1289, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1481, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1845, in _execute_context
    self._handle_dbapi_exception(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2026, in _handle_dbapi_exception
    util.raise_(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1782, in _execute_context
    self.dialect.do_executemany(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 729, in do_executemany
    cursor.executemany(statement, parameters)
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: skus.code
[SQL: INSERT INTO skus (code, formatted_code, product_index) VALUES (?, ?, ?)]
[parameters: (('0052135', '005-2135-8', 1), ('0072262', '007-2262-0', 1), ('0052133', '005-2133-2', 1), ('0052134', '005-2134-0', 1), ('0072260', '007-2260-4', 1), ('0052131', '005-2131-6', 1), ('0052132', '005-2132-4', 1), ('0072266', '007-2266-2', 1)  ... displaying 10 of 315833 total bound parameter sets ...  ('0438686', '043-8686-8', 127037), ('0529514', '052-9514-0', 127038))]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
```

jgalar commented:
I'm looking into why, but at least I thought I'd keep the error around :)

jgalar commented:
Adding a bit of debug code to the migration script, I found that, indeed, some skus are repeated:

```
>>> Reading product list, generating SKU list
--- Found duplicate skus ---
already present: {'index': 107568, 'name': 'Garrison Single Deadbolt Door Lock', 'product_code': '0465688P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 36, 18, 974584), 'url': '/en/pdp/garrison-single-deadbolt-door-lock-0465688p.html', 'sku_codes': '46-5690-8|46-5688-6'}
new:             {'index': 126557, 'name': 'Garrison Single Deadbolt Door Lock, Oxidized Bronze', 'product_code': '0465690P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 36, 39, 492726), 'url': '/en/pdp/garrison-single-deadbolt-door-lock-oxidized-bronze-0465690p.html', 'sku_codes': '46-5690-8'}
--- Found duplicate skus ---
already present: {'index': 107980, 'name': 'Mastercraft Aviation Snips, 10-in', 'product_code': '0585820P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 35, 24, 753810), 'url': '/en/pdp/mastercraft-aviation-snips-10-in-0585820p.html', 'sku_codes': '58-5822-8|58-5820-2|58-5821-0'}
new:             {'index': 126975, 'name': 'Mastercraft Aviation Snips, Left-Cut, 10-in', 'product_code': '0585821P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 11, 1, 45, 0, 917328), 'url': '/en/pdp/mastercraft-aviation-snips-left-cut-10-in-0585821p.html', 'sku_codes': '58-5821-0'}
--- Found duplicate skus ---
already present: {'index': 107980, 'name': 'Mastercraft Aviation Snips, 10-in', 'product_code': '0585820P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 35, 24, 753810), 'url': '/en/pdp/mastercraft-aviation-snips-10-in-0585820p.html', 'sku_codes': '58-5822-8|58-5820-2|58-5821-0'}
new:             {'index': 126977, 'name': 'Mastercraft Aviation Snips, Right-Cut, 10-in', 'product_code': '0585822P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 35, 38, 321139), 'url': '/en/pdp/mastercraft-aviation-snips-right-cut-10-in-0585822p.html', 'sku_codes': '58-5822-8'}
--- Found duplicate skus ---
already present: {'index': 105950, 'name': 'MAXIMUM 12-Piece Clinch Long Beam Wrench Set', 'product_code': '0580252P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 35, 17, 674359), 'url': '/en/pdp/maximum-12-piece-clinch-long-beam-wrench-set-0580252p.html', 'sku_codes': '58-0253-2|58-0252-4'}
new:             {'index': 126982, 'name': 'MAXIMUM Clinch Long Beam Wrench Set, Metric, 12-pc', 'product_code': '0580253P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 35, 35, 690990), 'url': '/en/pdp/maximum-clinch-long-beam-wrench-set-metric-12-pc-0580253p.html', 'sku_codes': '58-0253-2'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127008, 'name': 'Testors Craft Acrylic Matte Paint, Black, 59-mL', 'product_code': '0488009P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 182037), 'url': '/en/pdp/testors-craft-acrylic-matte-paint-black-59-ml-0488009p.html', 'sku_codes': '48-8009-2'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127009, 'name': 'Testors Craft Acrylic Matte Paint, Sunshine, 59-mL', 'product_code': '0488013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 179874), 'url': '/en/pdp/testors-craft-acrylic-matte-paint-sunshine-59-ml-0488013p.html', 'sku_codes': '48-8013-0'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127010, 'name': 'Testors Craft Acrylic Matte Paint, White, 59-mL', 'product_code': '0488010P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 184230), 'url': '/en/pdp/testors-craft-acrylic-matte-paint-white-59-ml-0488010p.html', 'sku_codes': '48-8010-6'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127011, 'name': 'Testors Craft Acrylic Paint, Fluorescent Pink, 59-mL', 'product_code': '0488030P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 177677), 'url': '/en/pdp/testors-craft-acrylic-paint-fluorescent-pink-59-ml-0488030p.html', 'sku_codes': '48-8030-8'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127012, 'name': 'Testors Craft Acrylic Metallic Paint, Black, 59-mL', 'product_code': '0488026P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 186451), 'url': '/en/pdp/testors-craft-acrylic-metallic-paint-black-59-ml-0488026p.html', 'sku_codes': '48-8026-0'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127013, 'name': 'Testors Craft Acrylic Matte Paint, Turquoise, 59-mL', 'product_code': '0488015P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 188621), 'url': '/en/pdp/testors-craft-acrylic-matte-paint-turquoise-59-ml-0488015p.html', 'sku_codes': '48-8015-6'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127014, 'name': 'Testors Craft Acrylic Glitter Burst Paint, Gold, 59-mL', 'product_code': '0488027P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 171052), 'url': '/en/pdp/testors-craft-acrylic-glitter-burst-paint-gold-59-ml-0488027p.html', 'sku_codes': '48-8027-8'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127015, 'name': 'Testors Craft Acrylic Paint, Fluorescent Orange, 59-mL', 'product_code': '0488029P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 173255), 'url': '/en/pdp/testors-craft-acrylic-paint-fluorescent-orange-59-ml-0488029p.html', 'sku_codes': '48-8029-4'}
--- Found duplicate skus ---
already present: {'index': 60300, 'name': 'Testors Craft Acrylic Craft Paint, 59-mL', 'product_code': '0488032P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 49, 949321), 'url': '/en/pdp/testors-craft-acrylic-craft-paint-59-ml-0488032p.html', 'sku_codes': '48-8009-2|48-8029-4|48-8015-6|48-8026-0|48-8027-8|48-8013-0|48-8034-0|48-8010-6|48-8032-4|48-8030-8'}
new:             {'index': 127016, 'name': 'Testors Craft Acrylic Matte Paint, Hazey, 59-mL', 'product_code': '0488034P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 54, 53, 168834), 'url': '/en/pdp/testors-craft-acrylic-matte-paint-hazey-59-ml-0488034p.html', 'sku_codes': '48-8034-0'}
--- Found duplicate skus ---
already present: {'index': 88304, 'name': 'Sherwood Hockey Throat Collar', 'product_code': '0832706P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 24, 40, 29872), 'url': '/en/pdp/sherwood-hockey-throat-collar-0832706p.html', 'sku_codes': '83-2705-4|83-2706-2'}
new:             {'index': 127018, 'name': 'Sherwood Hockey Throat Collar, Junior', 'product_code': '0832705P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 15, 19, 20, 46, 61249), 'url': '/en/pdp/sherwood-hockey-throat-collar-junior-0832705p.html', 'sku_codes': '83-2705-4'}
--- Found duplicate skus ---
already present: {'index': 89380, 'name': 'Warrior Street Hockey Glove & Blocker Set, Intermediate', 'product_code': '0838485P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 24, 40, 998921), 'url': '/en/pdp/warrior-street-hockey-glove-blocker-set-intermediate-0838485p.html', 'sku_codes': '83-8486-8|83-8485-0'}
new:             {'index': 127019, 'name': 'Warrior Street Hockey Glove & Blocker Set, Intermediate, Right', 'product_code': '0838486P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 25, 19, 891354), 'url': '/en/pdp/warrior-street-hockey-glove-blocker-set-intermediate-right-0838486p.html', 'sku_codes': '83-8486-8'}
--- Found duplicate skus ---
already present: {'index': 88424, 'name': 'Sherwood Hockey Neck Guard with Bib', 'product_code': '0832704P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 24, 39, 994747), 'url': '/en/pdp/sherwood-hockey-neck-guard-with-bib-0832704p.html', 'sku_codes': '83-2703-8|83-2704-6'}
new:             {'index': 127020, 'name': 'Sherwood Hockey Neck Guard with Bib, Junior', 'product_code': '0832703P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 25, 20, 337185), 'url': '/en/pdp/sherwood-hockey-neck-guard-with-bib-junior-0832703p.html', 'sku_codes': '83-2703-8'}
--- Found duplicate skus ---
already present: {'index': 63972, 'name': 'Bauer N8 Nechtech Hockey Neck Guard with Bib', 'product_code': '0832448P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 56, 24, 393042), 'url': '/en/pdp/bauer-n8-nechtech-hockey-neck-guard-with-bib-0832448p.html', 'sku_codes': '83-2447-8|83-2448-6'}
new:             {'index': 127021, 'name': 'Bauer N8 Nechtech Hockey Neck Guard with Bib, Senior', 'product_code': '0832447P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 25, 19, 425062), 'url': '/en/pdp/bauer-n8-nechtech-hockey-neck-guard-with-bib-senior-0832447p.html', 'sku_codes': '83-2447-8'}
--- Found duplicate skus ---
already present: {'index': 88061, 'name': 'CCM Street Hockey Glove & Blocker, Junior', 'product_code': '0839210P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 24, 35, 883958), 'url': '/en/pdp/ccm-street-hockey-glove-blocker-junior-0839210p.html', 'sku_codes': '83-9210-8|83-9211-6'}
new:             {'index': 127028, 'name': 'CCM Street Hockey Glove & Blocker, Junior, Right', 'product_code': '0839211P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 16, 1, 25, 19, 889678), 'url': '/en/pdp/ccm-street-hockey-glove-blocker-junior-right-0839211p.html', 'sku_codes': '83-9211-6'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 62, 'name': 'Michelin Defender T+H Tire', 'product_code': '0051014P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 313531), 'url': '/en/tires/pdp/michelin-defender-t-h-tire-0051014p.html', 'sku_codes': '408-2010-6|408-2012-2|408-2011-4|308-6480-0|408-2013-0|7-8521-8|6-1013-2|408-1995-8|6-1016-6|408-2006-8|408-1996-6|408-2009-2|408-1997-4|408-2008-4|408-1998-2|408-1999-0|408-2001-8|408-2000-0|408-2003-4|408-2005-0|408-1994-0|408-2004-2|7-8519-6|109-0098-2|5-1830-8|7-1029-4|7-8518-8|5-1828-6|6-1021-2|108-0098-4'}
new:             {'index': 127032, 'name': 'Michelin Defender T+H A/S Tire', 'product_code': '0061013P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 0, 16, 35, 310644), 'url': '/en/tires/pdp/michelin-defender-t-h-a-s-tire-0061013p.html', 'sku_codes': '6-1041-4|408-2011-4|408-2000-0|6-1013-2|408-2013-0|6-1021-2|408-1995-8|408-2006-8|7-8519-6|408-2009-2|7-1029-4|7-8518-8'}
--- Found duplicate skus ---
already present: {'index': 73458, 'name': 'Natural Salt Lamp', 'product_code': '0529515P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 1, 0, 23, 53926), 'url': '/en/pdp/natural-salt-lamp-0529515p.html', 'sku_codes': '52-9515-8|52-9514-0'}
new:             {'index': 127038, 'name': 'Natural Salt Lamp, 7-in', 'product_code': '0529514P', 'is_in_clearance': False, 'last_listed': datetime.datetime(2022, 7, 17, 1, 1, 6, 562189), 'url': '/en/pdp/natural-salt-lamp-7-in-0529514p.html', 'sku_codes': '52-9514-0'}
```

jgalar commented:
I have fixed the migration script, but I'm getting this error when scraping the inventory:

```
Scraping inventory  [----------------------------------#-]  2627  Champion Platinum Spark Plug,...
Traceback (most recent call last):
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: skus.code

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/jgalar/Projects/CanadianTracker/src/canadiantracker/scraper.py", line 137, in scrape_inventory
    repository.add_product_listing_entry(product_listing)
  File "/home/jgalar/Projects/CanadianTracker/src/canadiantracker/storage.py", line 251, in add_product_listing_entry
    .first()
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2810, in first
    return self.limit(1)._iter().first()
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2894, in _iter
    result = self.session.execute(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1640, in execute
    ) = compile_state_cls.orm_pre_session_exec(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 318, in orm_pre_session_exec
    session._autoflush()
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2237, in _autoflush
    util.raise_(e, with_traceback=sys.exc_info()[2])
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2226, in _autoflush
    self.flush()
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3363, in flush
    self._flush(objects)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3502, in _flush
    with util.safe_reraise():
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3463, in _flush
    flush_context.execute()
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 244, in save_obj
    _emit_insert_statements(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 1221, in _emit_insert_statements
    result = connection._execute_20(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1614, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1481, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1845, in _execute_context
    self._handle_dbapi_exception(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2026, in _handle_dbapi_exception
    util.raise_(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(sqlite3.IntegrityError) UNIQUE constraint failed: skus.code
[SQL: INSERT INTO skus (code, formatted_code, product_index) VALUES (?, ?, ?)]
[parameters: ('0225128', '022-5128-6', 2975)]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
Exception ignored in: <function _SQLite3ProductRepository.__del__ at 0x7fdde423aa70>
Traceback (most recent call last):
  File "/home/jgalar/Projects/CanadianTracker/src/canadiantracker/storage.py", line 206, in __del__
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1431, in commit
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 827, in commit
  File "/home/jgalar/.cache/pypoetry/virtualenvs/canadiantracker-p8QAoAzB-py3.10/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 601, in _assert_active
sqlalchemy.exc.PendingRollbackError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(sqlite3.IntegrityError) UNIQUE constraint failed: skus.code
[SQL: INSERT INTO skus (code, formatted_code, product_index) VALUES (?, ?, ?)]
[parameters: ('0225128', '022-5128-6', 2975)]
```

jgalar committed:
Skip duplicate skus during database migration

Some skus are present both as stand-alone static products, but
also as part of a static product entry's list of skus.

Skip the duplicates to allow the migration to proceeed while
respecting the sku table's constraints.

Signed-off-by: Jérémie Galarneau <jeremie.galarneau@gmail.com>

jgalar committed:
Fix typos in storage comments

Signed-off-by: Jérémie Galarneau <jeremie.galarneau@gmail.com>

jgalar committed:
Handle sku migrations from one product entry to another

When listing the product inventory, the triangle API can return a sku
that was previously associated to a different product entry.

These kind of "migrations" of skus happen periodically when a sku is
re-parented to a different product (typically because it changed names).
In this case, edit the existing sku entry to point to the more recent
product listing entry.

Signed-off-by: Jérémie Galarneau <jeremie.galarneau@gmail.com>

simark commented on src/canadiantracker/storage.py:
```diff
@@ -283,7 +283,20 @@ def add_product_listing_entry(
+                logger.debug(
+                    f"  SKU {sku.code} is associated to a different product: previous product was '{stored_sku.product.name} ({stored_sku.product.code})', new product is '{entry.name} ({entry.code})'"
+                )
+                stored_sku.entry = entry
```
What's the `entry` in `stored_sku.entry`?

simark replied:
Wondering if it should be `stored_sku.product = entry`.

jgalar replied:
My friend, you are quite correct. Thanks!
